## Libraries

In [ ]:
#!pip install autogluon

In [ ]:
import pandas as pd

from autogluon.tabular import TabularDataset, TabularPredictor

## Data preparation

Read in data from prepared file.

In [ ]:
df = pd.read_csv('df_prepped.csv')
pd.set_option('display.max_columns', None)
df

Year Countries  Sand_1  Sand_2  Sand_3  Sand_4  Sand_5  Sand_6  Sand_7  \
0      2007    Angola      50      51      51      48      45      46      46   
1      2007    Angola      62      64      63      59      58      59      59   
2      2007    Angola      69      71      70      67      65      65      66   
3      2007    Angola      60      63      61      57      53      53      53   
4      2007    Angola      67      69      68      63      61      61      61   
...     ...       ...     ...     ...     ...     ...     ...     ...     ...   
32354  2016  Zimbabwe      73      75      74      69      66      66      66   
32355  2016  Zimbabwe      57      58      58      54      52      51      52   
32356  2016  Zimbabwe      55      57      56      54      52      51      51   
32357  2016  Zimbabwe      70      72      71      67      63      63      62   
32358  2016  Zimbabwe      62      64      63      60      59      60      60   

       Clay_1  Clay_2  Clay_3  Clay_4  Clay_5  Clay_6  Clay_7  OC_1  OC_2  \
0          37      35      36      39      42      42      42  0.52  0.23   
1          27      25      26      29      31      30      30  0.11  0.05   
2          19      16      18      21      24      24      23  0.09  0.06   
3          29      26      28      32      35      36      36  0.46  0.16   
4          22      19      21      25      28      28      29  0.15  0.09   
...       ...     ...     ...     ...     ...     ...     ...   ...   ...   
32354      20      18      20      24      27      27      27  0.10  0.08   
32355      28      26      26      31      33      34      33  0.14  0.12   
32356      30      28      30      32      34      35      35  0.53  0.41   
32357      17      14      15      20      24      24      24  0.14  0.09   
32358      24      22      24      27      29      28      28  0.08  0.06   

       OC_3  OC_4  OC_5  OC_6  OC_7  PAW_1  PAW_2  PAW_3  PAW_4  PAW_5  PAW_6  \
0      0.17  0.09  0.04  0.02  0.02   0.15   0.15   0.14   0.13   0.10   0.07   
1      0.07  0.04  0.02  0.02  0.01   0.11   0.10   0.10   0.09   0.07   0.07   
2      0.07  0.04  0.02  0.02  0.02   0.10   0.10   0.10   0.09   0.07   0.07   
3      0.14  0.08  0.05  0.04  0.03   0.12   0.13   0.12   0.12   0.11   0.10   
4      0.09  0.05  0.02  0.01  0.01   0.11   0.11   0.11   0.11   0.08   0.04   
...     ...   ...   ...   ...   ...    ...    ...    ...    ...    ...    ...   
32354  0.08  0.04  0.02  0.01  0.01   0.09   0.10   0.09   0.09   0.07   0.04   
32355  0.10  0.05  0.03  0.01  0.01   0.13   0.14   0.13   0.11   0.09   0.03   
32356  0.30  0.22  0.18  0.16  0.15   0.18   0.19   0.17   0.15   0.14   0.14   
32357  0.10  0.06  0.04  0.03  0.03   0.10   0.10   0.10   0.10   0.09   0.09   
32358  0.06  0.03  0.02  0.01  0.01   0.11   0.11   0.11   0.09   0.08   0.04   

       PAW_7  Y_maize_major           Farm  Sow_Maize_month_int  \
0       0.07       0.615357     104_Angola                    9   
1       0.03       0.257656      99_Angola                    9   
2       0.07       4.286831     108_Angola                    9   
3       0.09       0.700384     102_Angola                    9   
4       0.04       0.553450      43_Angola                    9   
...      ...            ...            ...                  ...   
32354   0.04       0.674080  3874_Zimbabwe                    5   
32355   0.03       0.707797  3875_Zimbabwe                    5   
32356   0.14       0.805844  3859_Zimbabwe                    5   
32357   0.08       0.595883  3871_Zimbabwe                    5   
32358   0.04       0.846773  3869_Zimbabwe                    5   

       Harvest_Maize_month_int  sow_to_harvest_months    pcp_mean   tmax_mean  \
0                            4                      7  114.212553  301.525359   
1                            4                      7   29.700994  304.262436   
2                            4                      7   20.730178  305.844360   
3      

For faster demo testing, create subsampling. Default is whole dataset.

In [ ]:
subsample_size = len(df)  # subsample subset of data for faster demo; full dataset (32K) fit takes appr 18 min, sample of 3K takes 1.5 min.
df = df.sample(n=subsample_size, random_state=0)

In [ ]:
df.shape

(32359, 46)

Additionally, `Countries` and `Farm` attributes as categorical are dropped.

In [ ]:
df = df.drop(['Countries','Farm'], axis=1)
df

Year  Sand_1  Sand_2  Sand_3  Sand_4  Sand_5  Sand_6  Sand_7  Clay_1  \
17129  2011      65      67      66      61      60      60      60      22   
8309   2009      40      40      39      39      39      40      43      38   
16831  2011      57      57      55      52      51      51      51      26   
27283  2015      80      83      82      80      78      77      78      13   
29705  2016      60      63      61      58      52      50      51      21   
...     ...     ...     ...     ...     ...     ...     ...     ...     ...   
13123  2010      53      53      52      50      49      50      50      23   
19648  2012      24      25      25      24      25      26      26      49   
9845   2009      46      47      46      44      42      42      43      33   
10799  2009      49      53      52      49      45      45      46      29   
2732   2007      36      37      37      34      34      34      35      36   

       Clay_2  Clay_3  Clay_4  Clay_5  Clay_6  Clay_7  OC_1  OC_2  OC_3  OC_4  \
17129      20      22      26      27      27      28  0.10  0.08  0.07  0.03   
8309       37      38      38      40      39      37  0.39  0.25  0.20  0.14   
16831      24      27      30      32      31      31  0.33  0.06  0.06  0.04   
27283      10      11      14      16      17      17  0.09  0.04  0.05  0.03   
29705      19      21      26      32      34      32  0.23  0.15  0.11  0.05   
...       ...     ...     ...     ...     ...     ...   ...   ...   ...   ...   
13123      22      23      24      25      25      24  0.29  0.11  0.12  0.09   
19648      47      48      50      50      50      49  0.51  0.29  0.20  0.12   
9845       32      33      36      38      38      38  0.64  0.62  0.47  0.41   
10799      26      28      31      35      36      35  0.28  0.18  0.13  0.07   
2732       36      37      39      40      39      39  0.37  0.13  0.11  0.06   

       OC_5  OC_6  OC_7  PAW_1  PAW_2  PAW_3  PAW_4  PAW_5  PAW_6  PAW_7  \
17129  0.01  0.01  0.01   0.11   0.12   0.11   0.09   0.04   0.04   0.04   
8309   0.13  0.11  0.11   0.12   0.11   0.11   0.11   0.10   0.09   0.10   
16831  0.02  0.01  0.01   0.12   0.11   0.11   0.09   0.07   0.03   0.03   
27283  0.02  0.01  0.01   0.08   0.07   0.07   0.07   0.07   0.04   0.04   
29705  0.02  0.01  0.01   0.15   0.15   0.15   0.13   0.10   0.05   0.05   
...     ...   ...   ...    ...    ...    ...    ...    ...    ...    ...   
13123  0.07  0.06  0.06   0.13   0.13   0.13   0.12   0.11   0.11   0.11   
19648  0.10  0.08  0.08   0.17   0.16   0.14   0.13   0.12   0.11   0.11   
9845   0.36  0.34  0.33   0.18   0.18   0.15   0.15   0.15   0.14   0.14   
10799  0.04  0.03  0.03   0.15   0.15   0.15   0.14   0.12   0.10   0.10   
2732   0.03  0.03  0.02   0.12   0.11   0.11   0.10   0.07   0.07   0.06   

       Y_maize_major  Sow_Maize_month_int  Harvest_Maize_month_int  \
17129       2.233478                   10                        5   
8309        0.306847                    3                       10   
16831       0.201362                    3                       12   
27283       1.808289                    1                        6   
29705       1.683214                    3                        8   
...              ...                  ...                      ...   
13123       0.611172                    6                       12   
19648       1.618134                    3                       12   
9845        0.941790                   11                        4   
10799       2.915128                    3                        8   
2732        0.339835                    9                        4   

       sow_to_harvest_months    pcp_mean   tmax_mean   tmin_mean  spi_mean  \
17129                      7   44.754863  302.130629  290.975318  0.396764   
8309                       7   32.951514  309.732613  296.418146 -0.019928   
16831                      9   19.898077  306.946552  296.737980 -1.814732   
27283              

Create train and test sets.

In [ ]:
# Separate a test set, the year 2016
df_test = df[df.Year == 2016]
df_train = df[df.Year < 2016]

In [ ]:
df_test.shape, df_train.shape

((2930, 44), (29429, 44))

## AutoGluon: fit, predict, feature importance

Reference: https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html

Dependent variable for this regression task is `Y_maize_major`.

In [ ]:
label = 'Y_maize_major'
df_train[label].describe()

count    29429.000000
mean         1.715768
std          1.548261
min          0.000000
25%          0.790341
50%          1.227406
75%          2.168698
max         14.542289
Name: Y_maize_major, dtype: float64

### Fit and predict on RMSE

Initialize and fit AutoGluon TabularPredictor. As two metrics - RMSE and MAE - are used in this project, these need to be fitted separately. Default metric for AutoGluon is RMSE.

In [ ]:
# takes about 18 minutes
metric = 'root_mean_squared_error' # rmse is default metric

predictor_rmse = TabularPredictor(label=label, eval_metric=metric).fit(df_train)

No path specified. Models will be saved in: "AutogluonModels/ag-20231214_162101"
No presets specified! To achieve strong results with AutoGluon, it is recommended to use the available presets.
	Recommended Presets (For more details refer to https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):
	presets='best_quality'   : Maximize accuracy. Default time_limit=3600.
	presets='high_quality'   : Strong accuracy with fast inference speed. Default time_limit=3600.
	presets='good_quality'   : Good accuracy with very fast inference speed. Default time_limit=3600.
	presets='medium_quality' : Fast training time, ideal for initial prototyping.
Beginning AutoGluon training ...
AutoGluon will save models to "AutogluonModels/ag-20231214_162101"
=================== System Info ===================
AutoGluon Version:  1.0.0
Python Version:     3.10.12
Operating System:   Linux
Platform Machine:   x86_64
Platform Version:   #1 SMP PREEMPT_DYNAMIC Sat Nov 18 15:31:17 UTC 2023
C

[1000]	valid_set's rmse: 0.242437
[2000]	valid_set's rmse: 0.231508
[3000]	valid_set's rmse: 0.228415
[4000]	valid_set's rmse: 0.227292
[5000]	valid_set's rmse: 0.226207
[6000]	valid_set's rmse: 0.225717
[7000]	valid_set's rmse: 0.225063
[8000]	valid_set's rmse: 0.225075
[9000]	valid_set's rmse: 0.224978
[10000]	valid_set's rmse: 0.22488


	-0.2248	 = Validation score   (-root_mean_squared_error)
	48.35s	 = Training   runtime
	2.3s	 = Validation runtime
Fitting model: LightGBM ...


[1000]	valid_set's rmse: 0.218599


	-0.2154	 = Validation score   (-root_mean_squared_error)
	13.16s	 = Training   runtime
	0.31s	 = Validation runtime
Fitting model: RandomForestMSE ...
	-0.2636	 = Validation score   (-root_mean_squared_error)
	275.37s	 = Training   runtime
	0.6s	 = Validation runtime
Fitting model: CatBoost ...
	-0.1986	 = Validation score   (-root_mean_squared_error)
	228.23s	 = Training   runtime
	0.03s	 = Validation runtime
Fitting model: ExtraTreesMSE ...
	-0.2571	 = Validation score   (-root_mean_squared_error)
	68.19s	 = Training   runtime
	0.32s	 = Validation runtime
Fitting model: NeuralNetFastAI ...
	-0.2323	 = Validation score   (-root_mean_squared_error)
	32.19s	 = Training   runtime
	0.05s	 = Validation runtime
Fitting model: XGBoost ...
	-0.2217	 = Validation score   (-root_mean_squared_error)
	23.53s	 = Training   runtime
	0.41s	 = Validation runtime
Fitting model: NeuralNetTorch ...
	-0.2595	 = Validation score   (-root_mean_squared_error)
	100.67s	 = Training   runtime
	0.03s	 = Valida

[1000]	valid_set's rmse: 0.223592
[2000]	valid_set's rmse: 0.222284
[3000]	valid_set's rmse: 0.221985
[4000]	valid_set's rmse: 0.22198
[5000]	valid_set's rmse: 0.221929
[6000]	valid_set's rmse: 0.22193


	-0.2219	 = Validation score   (-root_mean_squared_error)
	72.23s	 = Training   runtime
	2.16s	 = Validation runtime
Fitting model: WeightedEnsemble_L2 ...
	Ensemble Weights: {'CatBoost': 0.763, 'NeuralNetFastAI': 0.118, 'NeuralNetTorch': 0.053, 'LightGBMLarge': 0.053, 'LightGBM': 0.013}
	-0.197	 = Validation score   (-root_mean_squared_error)
	0.31s	 = Training   runtime
	0.0s	 = Validation runtime
AutoGluon training complete, total runtime = 886.58s ... Best model: "WeightedEnsemble_L2"
TabularPredictor saved. To load, use: predictor = TabularPredictor.load("AutogluonModels/ag-20231214_162101")


Now our trained model is used to make predictions on the new data:

In [ ]:
y_pred_rmse = predictor_rmse.predict(df_test.drop(columns=[label]))
y_pred_rmse.head()

29705    1.671171
30082    0.670515
32243    3.635991
30623    3.044881
31434    0.552511
Name: Y_maize_major, dtype: float32

Next, we can evaluate the predictor on the (labeled) test data:

In [ ]:
predictor_rmse.evaluate(df_test, silent=True)

{'root_mean_squared_error': -0.3801093725544506,
 'mean_squared_error': -0.14448313510373811,
 'mean_absolute_error': -0.2389560943166964,
 'r2': 0.9458992344944107,
 'pearsonr': 0.9738149407906912,
 'median_absolute_error': -0.14647175370378873}

For evaluation metrics where higher values are worse (like RMSE), AutoGluon will flip their sign and print them as negative values during training (as it internally assumes higher values are better).

Each model is evaluated individually, in this case based on RMSE.

In [ ]:
predictor_rmse.leaderboard(df_test)

model  score_test  score_val              eval_metric  \
0         ExtraTreesMSE   -0.312985  -0.257074  root_mean_squared_error   
1            LightGBMXT   -0.349807  -0.224799  root_mean_squared_error   
2       RandomForestMSE   -0.356448  -0.263565  root_mean_squared_error   
3   WeightedEnsemble_L2   -0.380109  -0.196951  root_mean_squared_error   
4         LightGBMLarge   -0.385576  -0.221926  root_mean_squared_error   
5               XGBoost   -0.386810  -0.221728  root_mean_squared_error   
6              LightGBM   -0.386854  -0.215393  root_mean_squared_error   
7              CatBoost   -0.391063  -0.198636  root_mean_squared_error   
8        NeuralNetTorch   -0.397514  -0.259467  root_mean_squared_error   
9       NeuralNetFastAI   -0.418125  -0.232311  root_mean_squared_error   
10       KNeighborsDist   -0.688649  -0.646700  root_mean_squared_error   
11       KNeighborsUnif   -0.710672  -0.667693  root_mean_squared_error   

    pred_time_test  pred_time_val    fit_time  pred_time_test_marginal  \
0         0.650309       0.320880   68.190084                 0.650309   
1         3.643582       2.300006   48.346956                 3.643582   
2         1.558896       0.595390  275.373967                 1.558896   
3         3.419978       2.586540  446.789228                 0.012492   
4         2.629050       2.159232   72.226476                 2.629050   
5         0.566243       0.408810   23.527739                 0.566243   
6         0.549800       0.314956   13.164969                 0.549800   
7         0.058407       0.028895  228.228576                 0.058407   
8         0.040949       0.031129  100.671959                 0.040949   
9         0.129280       0.051833   32.189833                 0.129280   
10        0.534786       0.450071    0.043199                 0.534786   
11        0.532984       0.541061    2.915728                 0.532984   

    pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  \
0                 0.320880          68.190084            1       True   
1                 2.300006          48.346956            1       True   
2                 0.595390         275.373967            1       True   
3                 0.000494           0.307414            2       True   
4                 2.159232          72.226476            1       True   
5                 0.408810          23.527739            1       True   
6                 0.314956          13.164969            1       True   
7                 0.028895         228.228576            1       True   
8                 0.031129         100.671959            1       True   
9                 0.051833          32.189833            1       True   
10                0.450071           0.043199            1       True   
11                0.541061           2.915728            1       True   

    fit_order  
0           7  
1           3  
2           5  
3          12  
4          11  
5           9  
6           4  
7           6  
8          10  
9           8  
10          2  
11          1

Best model is ExtraTreesMSE with RMSE on test set of 0.313.

For the whole model, RMSE on test data is 0.3801 and MAE 0.239.

### Fit and predict on MAE

Same procedure is now followed by using mean absolute error as metric. First fit on train data:

In [ ]:
metric = 'mean_absolute_error'
predictor_mae = TabularPredictor(label=label, eval_metric=metric).fit(df_train)

No path specified. Models will be saved in: "AutogluonModels/ag-20231214_163605"
No presets specified! To achieve strong results with AutoGluon, it is recommended to use the available presets.
	Recommended Presets (For more details refer to https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):
	presets='best_quality'   : Maximize accuracy. Default time_limit=3600.
	presets='high_quality'   : Strong accuracy with fast inference speed. Default time_limit=3600.
	presets='good_quality'   : Good accuracy with very fast inference speed. Default time_limit=3600.
	presets='medium_quality' : Fast training time, ideal for initial prototyping.
Beginning AutoGluon training ...
AutoGluon will save models to "AutogluonModels/ag-20231214_163605"
=================== System Info ===================
AutoGluon Version:  1.0.0
Python Version:     3.10.12
Operating System:   Linux
Platform Machine:   x86_64
Platform Version:   #1 SMP PREEMPT_DYNAMIC Sat Nov 18 15:31:17 UTC 2023
C

[1000]	valid_set's l1: 0.141738
[2000]	valid_set's l1: 0.135217
[3000]	valid_set's l1: 0.132685
[4000]	valid_set's l1: 0.131697
[5000]	valid_set's l1: 0.130956
[6000]	valid_set's l1: 0.130524
[7000]	valid_set's l1: 0.130052
[8000]	valid_set's l1: 0.129936
[9000]	valid_set's l1: 0.129876
[10000]	valid_set's l1: 0.129622


	-0.1296	 = Validation score   (-mean_absolute_error)
	47.39s	 = Training   runtime
	2.87s	 = Validation runtime
Fitting model: LightGBM ...


[1000]	valid_set's l1: 0.130815
[2000]	valid_set's l1: 0.126751
[3000]	valid_set's l1: 0.12566
[4000]	valid_set's l1: 0.12516
[5000]	valid_set's l1: 0.124813
[6000]	valid_set's l1: 0.124749


	-0.1247	 = Validation score   (-mean_absolute_error)
	34.36s	 = Training   runtime
	1.1s	 = Validation runtime
Fitting model: RandomForestMSE ...
	-0.1517	 = Validation score   (-mean_absolute_error)
	281.81s	 = Training   runtime
	0.34s	 = Validation runtime
Fitting model: CatBoost ...
	-0.1188	 = Validation score   (-mean_absolute_error)
	295.84s	 = Training   runtime
	0.03s	 = Validation runtime
Fitting model: ExtraTreesMSE ...
	-0.1507	 = Validation score   (-mean_absolute_error)
	70.81s	 = Training   runtime
	0.33s	 = Validation runtime
Fitting model: NeuralNetFastAI ...
	-0.1433	 = Validation score   (-mean_absolute_error)
	31.82s	 = Training   runtime
	0.04s	 = Validation runtime
Fitting model: XGBoost ...
	-0.1285	 = Validation score   (-mean_absolute_error)
	23.83s	 = Training   runtime
	0.31s	 = Validation runtime
Fitting model: NeuralNetTorch ...
	-0.1691	 = Validation score   (-mean_absolute_error)
	57.79s	 = Training   runtime
	0.04s	 = Validation runtime
Fitting model: L

[1000]	valid_set's l1: 0.126124
[2000]	valid_set's l1: 0.124174
[3000]	valid_set's l1: 0.123779
[4000]	valid_set's l1: 0.123682
[5000]	valid_set's l1: 0.1236
[6000]	valid_set's l1: 0.123569
[7000]	valid_set's l1: 0.123566
[8000]	valid_set's l1: 0.12356
[9000]	valid_set's l1: 0.123557
[10000]	valid_set's l1: 0.123558


	-0.1236	 = Validation score   (-mean_absolute_error)
	103.16s	 = Training   runtime
	4.04s	 = Validation runtime
Fitting model: WeightedEnsemble_L2 ...
	Ensemble Weights: {'CatBoost': 0.464, 'LightGBMLarge': 0.214, 'LightGBMXT': 0.125, 'LightGBM': 0.107, 'NeuralNetFastAI': 0.089}
	-0.1159	 = Validation score   (-mean_absolute_error)
	0.57s	 = Training   runtime
	0.0s	 = Validation runtime
AutoGluon training complete, total runtime = 975.19s ... Best model: "WeightedEnsemble_L2"
TabularPredictor saved. To load, use: predictor = TabularPredictor.load("AutogluonModels/ag-20231214_163605")


Then make predictions:

In [ ]:
y_pred_mae = predictor_mae.predict(df_test.drop(columns=[label]))
y_pred_mae.head()

29705    1.622712
30082    0.669600
32243    3.664353
30623    3.016555
31434    0.569303
Name: Y_maize_major, dtype: float32

Then evaluation on labeled test data:

In [ ]:
predictor_mae.evaluate(df_test, silent=True)

{'mean_absolute_error': -0.23112485545675243,
 'root_mean_squared_error': -0.3660497589452939,
 'mean_squared_error': -0.13399242602390776,
 'r2': 0.9498274119353816,
 'pearsonr': 0.9761698412429389,
 'median_absolute_error': -0.1415425985928803}

And then model ranking based on MAE:

In [ ]:
predictor_mae.leaderboard(df_test)

model  score_test  score_val          eval_metric  \
0         ExtraTreesMSE   -0.196717  -0.150729  mean_absolute_error   
1        NeuralNetTorch   -0.212474  -0.169121  mean_absolute_error   
2       RandomForestMSE   -0.215526  -0.151741  mean_absolute_error   
3            LightGBMXT   -0.227660  -0.129598  mean_absolute_error   
4   WeightedEnsemble_L2   -0.231125  -0.115926  mean_absolute_error   
5         LightGBMLarge   -0.237375  -0.123557  mean_absolute_error   
6              LightGBM   -0.241110  -0.124694  mean_absolute_error   
7               XGBoost   -0.242689  -0.128455  mean_absolute_error   
8              CatBoost   -0.246546  -0.118787  mean_absolute_error   
9       NeuralNetFastAI   -0.260539  -0.143339  mean_absolute_error   
10       KNeighborsDist   -0.435562  -0.420583  mean_absolute_error   
11       KNeighborsUnif   -0.452493  -0.436629  mean_absolute_error   

    pred_time_test  pred_time_val    fit_time  pred_time_test_marginal  \
0         0.637584       0.331300   70.809156                 0.637584   
1         0.038663       0.042961   57.789162                 0.038663   
2         1.203003       0.340592  281.813240                 1.203003   
3         3.596737       2.869782   47.386438                 3.596737   
4        10.678272       8.084863  513.136014                 0.023727   
5         5.479149       4.043400  103.156404                 5.479149   
6         1.378310       1.103882   34.361890                 1.378310   
7         0.476219       0.307049   23.827204                 0.476219   
8         0.049422       0.029467  295.840356                 0.049422   
9         0.150928       0.037649   31.819860                 0.150928   
10        0.851813       0.462522    0.038596                 0.851813   
11        0.848912       0.439755    0.035016                 0.848912   

    pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  \
0                 0.331300          70.809156            1       True   
1                 0.042961          57.789162            1       True   
2                 0.340592         281.813240            1       True   
3                 2.869782          47.386438            1       True   
4                 0.000683           0.571065            2       True   
5                 4.043400         103.156404            1       True   
6                 1.103882          34.361890            1       True   
7                 0.307049          23.827204            1       True   
8                 0.029467         295.840356            1       True   
9                 0.037649          31.819860            1       True   
10                0.462522           0.038596            1       True   
11                0.439755           0.035016            1       True   

    fit_order  
0           7  
1          10  
2           5  
3           3  
4          12  
5          11  
6           4  
7           9  
8           6  
9           8  
10          2  
11          1

Best model based on MAE is also ExtraTreesMSE with MAE on test data of 0.197.

### Feature importance

To better understand our trained predictor, we can estimate the overall importance of each feature via `feature_importance()`. First, model fitted based on RMSE:

In [ ]:
predictor_rmse.feature_importance(df_test) #14 minutes

Computing feature importance via permutation shuffling for 43 features using 2930 rows with 5 shuffle sets...
	844.21s	= Expected runtime (168.84s per shuffle set)
	572.95s	= Actual runtime (Completed 5 of 5 shuffle sets)


importance    stddev       p_value  n  p99_high  \
maize_lag-1                0.937821  0.011153  2.400145e-09  5  0.960786   
maize_lag-3                0.456286  0.012109  5.946619e-08  5  0.481218   
maize_lag-2                0.292626  0.011871  3.242670e-07  5  0.317068   
tmax_mean                  0.032987  0.003584  1.645763e-05  5  0.040366   
tmin_mean                  0.012222  0.000904  3.569271e-06  5  0.014084   
pcp_mean_lag-1             0.004192  0.001268  8.916833e-04  5  0.006802   
spi_mean_lag-1             0.003993  0.002890  1.828966e-02  5  0.009943   
pcp_mean                   0.003473  0.004810  9.083250e-02  5  0.013377   
OC_5                       0.001265  0.000536  3.087039e-03  5  0.002368   
PAW_6                      0.000845  0.000082  1.056331e-05  5  0.001014   
OC_6                       0.000491  0.000257  6.488405e-03  5  0.001021   
PAW_5                      0.000395  0.000172  3.382178e-03  5  0.000749   
spi_mean                   0.000357  0.002355  3.758744e-01  5  0.005206   
PAW_7                      0.000201  0.000177  3.160222e-02  5  0.000565   
Year                       0.000000  0.000000  5.000000e-01  5  0.000000   
Clay_3                    -0.000151  0.000425  7.648224e-01  5  0.000723   
Harvest_Maize_month_int   -0.000200  0.003210  5.519411e-01  5  0.006411   
Sand_4                    -0.000766  0.000435  9.914881e-01  5  0.000130   
OC_2                      -0.001799  0.000550  9.990690e-01  5 -0.000666   
OC_1                      -0.001899  0.000392  9.997933e-01  5 -0.001091   
PAW_4                     -0.002122  0.000595  9.993321e-01  5 -0.000898   
PAW_3                     -0.002380  0.000300  9.999703e-01  5 -0.001762   
tmax_mean_lag-1           -0.002387  0.002291  9.598715e-01  5  0.002330   
OC_4                      -0.002398  0.000413  9.998986e-01  5 -0.001548   
PAW_2                     -0.002608  0.000315  9.999748e-01  5 -0.001959   
Sand_2                    -0.002647  0.000459  9.998956e-01  5 -0.001701   
OC_3                      -0.002681  0.000576  9.997598e-01  5 -0.001495   
PAW_1                     -0.002796  0.000210  9.999962e-01  5 -0.002363   
Clay_4                    -0.003220  0.000315  9.999892e-01  5 -0.002572   
OC_7                      -0.003332  0.000179  9.999990e-01  5 -0.002963   
Clay_7                    -0.003531  0.000602  9.999026e-01  5 -0.002292   
Sand_5                    -0.004069  0.000912  9.997159e-01  5 -0.002190   
Clay_5                    -0.004159  0.000159  9.999997e-01  5 -0.003831   
Sand_3                    -0.004214  0.000666  9.999277e-01  5 -0.002843   
Clay_1                    -0.004778  0.000622  9.999663e-01  5 -0.003497   
Sand_7                    -0.005026  0.000352  9.999971e-01  5 -0.004303   
Sand_6                    -0.005115  0.000318  9.999982e-01  5 -0.004460   
Clay_6                    -0.005235  0.000381  9.999967e-01  5 -0.004451   
Sand_1                    -0.005682  0.000692  9.999741e-01  5 -0.004256   
tmix_mean_lag-1           -0.006160  0.001703  9.993657e-01  5 -0.002654   
Clay_2                    -0.006680  0.001038  9.999324e-01  5 -0.004544   
Sow_Maize_month_int       -0.020224  0.000990  9.999993e-01  5 -0.018185   
sow_to_harvest_months     -0.026615  0.002284  9.999936e-01  5 -0.021912   

                          p99_low  
maize_lag-1              0.914856  
maize_lag-3              0.431353  
maize_lag-2              0.268184  
tmax_mean                0.025608  
tmin_mean                0.010360  
pcp_mean_lag-1           0.001582  
spi_mean_lag-1          -0.001957  
pcp_mean                -0.006430  
OC_5                     0.000162  
PAW_6                    0.000676  
OC_6                    -0.000039  
PAW_5                    0.000042  
spi_mean                -0.004492  
PAW_7                   -0.000162  
Year                     0.000000  
Clay_3                  -0.001026  
Harvest_Maize_month_int -0.006810  
Sand_4                  -0.00166

Then, model fitted based on MAE:

In [ ]:
predictor_mae.feature_importance(df_test)

Computing feature importance via permutation shuffling for 43 features using 2930 rows with 5 shuffle sets...
	5875.75s	= Expected runtime (1175.15s per shuffle set)
	2024.28s	= Actual runtime (Completed 5 of 5 shuffle sets)


importance    stddev       p_value  n  p99_high  \
maize_lag-1                0.601828  0.007900  3.561709e-09  5  0.618093   
maize_lag-2                0.151326  0.006006  2.971454e-07  5  0.163692   
maize_lag-3                0.141432  0.006413  5.059907e-07  5  0.154638   
tmax_mean                  0.019797  0.001117  1.209326e-06  5  0.022096   
tmin_mean                  0.006888  0.000515  3.708831e-06  5  0.007947   
pcp_mean_lag-1             0.003759  0.000715  1.493658e-04  5  0.005231   
pcp_mean                   0.003066  0.001853  1.041261e-02  5  0.006881   
spi_mean_lag-1             0.002298  0.001336  9.188547e-03  5  0.005050   
spi_mean                   0.001193  0.001281  5.281034e-02  5  0.003830   
OC_5                       0.000435  0.000136  1.023225e-03  5  0.000716   
PAW_6                      0.000283  0.000076  5.835128e-04  5  0.000440   
tmax_mean_lag-1            0.000259  0.001261  3.349039e-01  5  0.002855   
tmix_mean_lag-1            0.000253  0.000675  2.241324e-01  5  0.001643   
PAW_5                      0.000225  0.000095  3.064932e-03  5  0.000422   
OC_6                       0.000164  0.000083  5.818467e-03  5  0.000335   
Year                       0.000000  0.000000  5.000000e-01  5  0.000000   
Sand_3                    -0.000335  0.000191  9.913501e-01  5  0.000059   
OC_7                      -0.000408  0.000042  9.999863e-01  5 -0.000321   
PAW_1                     -0.000412  0.000156  9.979614e-01  5 -0.000092   
PAW_7                     -0.000556  0.000112  9.998105e-01  5 -0.000325   
PAW_4                     -0.000568  0.000151  9.994477e-01  5 -0.000256   
OC_3                      -0.000681  0.000201  9.991846e-01  5 -0.000267   
PAW_3                     -0.000737  0.000098  9.999626e-01  5 -0.000534   
PAW_2                     -0.000739  0.000193  9.994919e-01  5 -0.000342   
Sand_2                    -0.000832  0.000149  9.998811e-01  5 -0.000525   
OC_4                      -0.000839  0.000116  9.999577e-01  5 -0.000601   
OC_1                      -0.001003  0.000366  9.982082e-01  5 -0.000250   
Sand_4                    -0.001154  0.000203  9.998890e-01  5 -0.000735   
Clay_4                    -0.001178  0.000204  9.998966e-01  5 -0.000759   
OC_2                      -0.001228  0.000123  9.999879e-01  5 -0.000974   
Clay_5                    -0.001321  0.000064  9.999993e-01  5 -0.001189   
Clay_7                    -0.001327  0.000193  9.999482e-01  5 -0.000930   
Clay_6                    -0.001423  0.000203  9.999521e-01  5 -0.001006   
Clay_2                    -0.001492  0.000406  9.994020e-01  5 -0.000656   
Clay_3                    -0.001637  0.000328  9.998159e-01  5 -0.000961   
Sand_5                    -0.001645  0.000219  9.999630e-01  5 -0.001193   
Clay_1                    -0.001888  0.000443  9.996620e-01  5 -0.000976   
Sand_6                    -0.002153  0.000198  9.999915e-01  5 -0.001746   
Sand_7                    -0.002530  0.000179  9.999970e-01  5 -0.002161   
Sand_1                    -0.002891  0.000299  9.999864e-01  5 -0.002275   
Harvest_Maize_month_int   -0.003285  0.000900  9.993857e-01  5 -0.001431   
Sow_Maize_month_int       -0.010369  0.000490  9.999994e-01  5 -0.009359   
sow_to_harvest_months     -0.017178  0.000675  9.999997e-01  5 -0.015788   

                          p99_low  
maize_lag-1              0.585562  
maize_lag-2              0.138959  
maize_lag-3              0.128227  
tmax_mean                0.017498  
tmin_mean                0.005828  
pcp_mean_lag-1           0.002288  
pcp_mean                -0.000749  
spi_mean_lag-1          -0.000454  
spi_mean                -0.001444  
OC_5                     0.000154  
PAW_6                    0.000125  
tmax_mean_lag-1         -0.002337  
tmix_mean_lag-1         -0.001136  
PAW_5                    0.000029  
OC_6                    -0.000007  
Year                     0.000000  
Sand_3                  -0.000729  
OC_7                    -0.00049